### ISSUE TITLE
Returning reborrowed reference in tuple causes unsoundness

### ISSUE THREAD

**Issue by dewert99** [xldenis/creusot#492]
Example running the `mlcfg` from
```
use creusot_contracts::*;

#[ensures(*(result.0) == *x)]
fn reborrow_tuple<T>(x: &mut T) -> (&mut T, u32) {
    (x, 32)
}

#[ensures(false)]
fn test() {
    let mut x = 5;
    let (res, _) = reborrow_tuple(&mut x);
    proof_assert!(^res == 5i32);
    *res = 10;
}
```
with `why3 prove $1 -L ../creusot/prelude/ -P z3 -P cvc4 -P alt-ergo -a split_vc -t 5`
gives the output
```
Warning, file "target/debug/my_test-rlib.mlcfg", line 99, characters 21-25:
unused variable self
File "...\src\fail.rs", line 3, characters 10-27:
Sub-goal Postcondition of goal reborrow_tuple'vc.
Prover result is: Valid (0.02s, 28 steps).

File "...\src\fail.rs", line 12, characters 18-30:
Sub-goal Assertion of goal test'vc.
Prover result is: Valid (0.03s, 77 steps).

File "...\src\fail.rs", line 8, characters 10-15:
Sub-goal Postcondition of goal test'vc.
Prover result is: Valid (0.03s, 57 steps).
```
Somehow the postcondition which only mentioned the past of `result.0` provided the prover information about its future, which allowed it to prove false when combined with the resolve statement.

**Comment by xldenis** [comment]
Thanks for the report! 

Looking at the generated code, the issue appears to really be with the splitting of the tuple rather than the function call (directly at least). 

The binding `let (res, _) = ... ` leads to the following code:

```
    res_2 <- (let (a, _) = _3 in a);
    assume { Resolve1.resolve _3 };
```

Here we should only resolve the second field of `_3` not the first component which was just moved out and is still alive.

**Comment by xldenis** [comment]
The best thing to do would be for the mlcfg to first replace the relevant field in `_3` with a symbolic value representing the fact it is now uninitialized

**Comment by xldenis** [comment]
Fixed in #493
